### ISSUE TITLE
macOS SDL 2.24.1: SDL_CloseAudioDevice locks forever, when omitted stuttering sound

### ISSUE THREAD

**Issue by slajerek** [libsdl-org/SDL#6377]
macOS Catalina 10.15.7
There's something messy in SDL 2.24.1 or I am missing some API change introduced in 2.24.1 that causes this.

SDL_CloseAudioDevice(1);

gives me endless lock. 

The code I am testing was working for more than two years already with previous versions of SDL2: https://github.com/slajerek/MTEngineSDL/blob/master/src/Engine/Audio/SND_SoundEngine.cpp

The case I am having here is to allow the user to switch active audio devices via a menu.

When I skip the device close at startup to avoid the endless lock and then open the selected audio device I have a very badly stuttered sound, and that does not seem to be a buffer underrun problem. The above code indeed quickly changes devices from default to selected by the user at startup, I amended that to avoid such a quick change, but that does not solve the problem. Whenever I have a custom audio device selected and playing, and then I would like to close it using SDL_CloseAudioDevice I have the endless lock.

All works perfectly fine in 2.0.22 (and many versions before), I reverted the SDL 2.24.1 upgrade for now.

Could you please advise what may be the problem here?

**Comment by sezero** [comment]
To make sure: Does 2.24.0 has the same issue as 2.24.1 ?

**Comment by slajerek** [comment]
@sezero I've just confirmed that the problem **does not** occur in **2.24.0**. Only the 2.24.1 is affected.

**Comment by sezero** [comment]
So it's the backports for https://github.com/libsdl-org/SDL/issues/6159, one of (or all of) the following:
https://github.com/libsdl-org/SDL/commit/d6e939ddb50dbe0c5c36ca57f52e2e678678a6d3
https://github.com/libsdl-org/SDL/commit/cedd8e11388d795a0e3d11cf2bb58041c2d8fc27
https://github.com/libsdl-org/SDL/commit/b0a2164b2c9c9d8b40ab4e9202d438d961002ff4

**Comment by sezero** [comment]
> @sezero I've just confirmed that the problem **does not** occur in **2.24.0**. Only the 2.24.1 is affected.

Out of curiosity, how does latest git version behave ??

> So it's the backports for #6159, one of (or all of) the following:
> [d6e939d](https://github.com/libsdl-org/SDL/commit/d6e939ddb50dbe0c5c36ca57f52e2e678678a6d3)
> [cedd8e1](https://github.com/libsdl-org/SDL/commit/cedd8e11388d795a0e3d11cf2bb58041c2d8fc27)
> [b0a2164](https://github.com/libsdl-org/SDL/commit/b0a2164b2c9c9d8b40ab4e9202d438d961002ff4)

@icculus: should we revert those ones from 2.24.x ?

**Comment by icculus** [comment]
Im going to look at this more closely before we revert anything.

**Comment by icculus** [comment]
Can you tell me: is this hanging because this LockMutex call...

```c
bool CSoundEngine::SetOutputAudioDevice(const char *deviceName)
{
	LOGM("CSoundEngine::SetOutputAudioDevice: deviceName=%s", deviceName ? deviceName : "NULL");
	
	this->LockMutex("CSoundEngine::SetOutputAudioDevice");
```

...is preventing the one in `playCallback` from completing?


> When I skip the device close at startup to avoid the endless lock and then open the selected audio device I have a very badly stuttered sound

This separate problem is probably because there are unexpectedly two threads calling playCallback, when the initial device close wasn't called, and they are taking turns getting the next chunk of the mixer output.

**Comment by slajerek** [comment]
@icculus No, it is handing on the SDL_CloseAudioDevice(1);

I've stopped the running process to see stack trace, please check this screenshot:

<img width="1104" alt="Screenshot 2022-10-18 at 22 10 00" src="https://user-images.githubusercontent.com/1021783/196533985-6ba8c679-a114-4822-a64d-6c015c66c734.png">

Note, this mechanism works fine on 2.24.0 (and previous versions when I implemented that, approximately since 2.0.10)

**Comment by icculus** [comment]
Is "BlackHole 16ch" using this thing? https://github.com/ExistentialAudio/BlackHole

**Comment by slajerek** [comment]
Yes, that's a virtual sound card for routing signals. But I've also just checked that it does not matter, changing any other device causes the same long delay (but indeed it completes, just takes a very long time).

**Comment by slajerek** [comment]
One more note, I have just confirmed that SDL_CloseAudioDevice and this mechanism do work fine on MS Windows with SDL 2.24.1

**Comment by icculus** [comment]
Okay, I'm going to try to get MTEngineSDL building over here and see I can reproduce it directly.

**Comment by slajerek** [comment]
Thanks, if you have any trouble compiling and running this please let me know. You probably need to add https://github.com/slajerek/RetroDebugger because the engine itself is a library and requires at least simple MT_API implementation: https://github.com/slajerek/MTEngineSDL/blob/master/src/MT_API.h

**Comment by icculus** [comment]
Okay, there's no chance I'm going to get this to build.  :)

My best guess at this point is that we moved the SDL_WaitThread() call above the AudioQueueDispose() call, which seems like the correct thing to do, but maybe it's causing some massive delay?

That change was b0a2164b2c9c9d8b40ab4e9202d438d961002ff4. If I just slide the AudioQueueDispose code above the `if (this->hidden->thread)` line, it still works here, and maybe fixes your issue?

(And if it fixes your issue, let's see if changing the 1 to 0 in that AudioQueueDispose call also works, as it currently throws away any audio data that has been sent to the queue but not yet reached the hardware.)

**Comment by slajerek** [comment]
@icculus Bingo! That fixed the problem. The code SDL_coreaudio.m around line 704 I have here is now:

```
    /* if callback fires again, feed silence; don't call into the app. */
    SDL_AtomicSet(&this->paused, 1);

    // moved AudioQueueDispose from below:
    if (this->hidden->audioQueue) {
        AudioQueueDispose(this->hidden->audioQueue, 0);
    }

    if (this->hidden->thread) {
        SDL_assert(SDL_AtomicGet(&this->shutdown) != 0);  /* should have been set by SDL_audio.c */
        SDL_WaitThread(this->hidden->thread, NULL);
    }
```

There's no huge lag anymore with this change and the problem is fixed. I checked both putting 0 or 1 and I hear no difference, both options work OK.

**Comment by icculus** [comment]
Ok, this is in revision control. Thanks for your patience as we tracked that down!

@sezero, this patch can go into 2.24.x.

(I _do_ want to say again that the mutex thing I mentioned before appears to be a legit race condition; if SetOutputAudioDevice locks that mutex, THEN playCallback in another thread tries to lock it, and THEN we try to close the device further into SetOutputAudioDevice, playCallback will never return because it's waiting on that mutex, and therefore SDL_CloseAudioDevice() will never return because it's waiting on playCallback to return. It's a very small window where things can go wrong, but it would be safer to put an SDL_PauseAudioDevice() call before the initial lock in SetOutputAudioDevice, which guarantees the callback is not running and will not run again.)

**Comment by slajerek** [comment]
@icculus Wow, thanks for the explanation, indeed that might cause the race condition you described. I have changed my code and now pause playback before calling SDL_CloseAudioDevice(). 

And of course, thank you for the fix!

**Comment by sezero** [comment]
> Ok, this is in revision control. Thanks for your patience as we tracked that down!
> 
> @sezero, this patch can go into 2.24.x.

Done.
